### PR TITLE
Add cluster-autoscaler

### PR DESCRIPTION
**Pull request by duboisph**
Adds all necessary components for setup of the cluster-autoscaler:
- kube2iam role & policy
- EC2 tags on worker nodes for ASG auto-discovery
- Generate [cluster-autoscaler helm values](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler)
- Grafana dashboard: https://grafana.test.skyscrape.rs/dashboard/db/kubernetes-cluster-autoscaler?refresh=1m&orgId=1&from=now-6h&to=now

I opted to enable the cluster-autoscaler by default on the workers instance group, since that makes the most sense for having an ASG anyway. This means of course that it's important to set correct min and max values on the ASG(s).

PR requirements:
- Related story: https://github.com/skyscrapers/engineering/issues/79
- Realted PR: https://github.com/skyscrapers/charts/pull/45
- Tested scale up and scale down events on our SkyscrapersTest cluster

### CODE REVIEW

File: base/iam.tf [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+      ],
+      "Effect": "Allow",
+      "Resource": "*"
+    }
+  ]
+}
+EOF
+}
+
+data "aws_region" "fluentd_region" {}
```
Comment (iuriaranda) [inline base/iam.tf]: You have another `aws_region` data provider to get the current region, line 5 of the `main.tf` file:

```
data "aws_region" "current" {}
```

File: base/main.tf [hunk 2]
```diff
@@ -350,3 +218,18 @@ resource "local_file" "helm_values_nginx-ingress_file" {
   content  = "${data.template_file.helm_values_nginx_ingress.rendered}"
   filename = "${path.cwd}/helm-values-nginx-ingress.yaml"
 }
+
+data "template_file" "helm_values_autoscaler" {
+  template = "${file("${path.module}/../templates/helm-values-autoscaler.tpl.yaml")}"
+
+  vars {
+    aws_region          = "${data.aws_region.current.name}"
+    cluster_name        = "${var.name}"
```
Comment (SamClinckspoor) [inline base/main.tf]: This is properly namespaces? eg no staging and prod clusters int he same account will affect eachother?

File: templates/kops-instancegroup-worker.tpl.yaml [hunk 3]
```diff
@@ -19,6 +19,8 @@ ${subnets}
     role: worker
     subnet: app
     access: private
+    k8s.io/cluster-autoscaler/enabled: ""
```
Comment (SamClinckspoor) [inline templates/kops-instancegroup-worker.tpl.yaml]: what does this do?

Comment (SamClinckspoor) [reply templates/kops-instancegroup-worker.tpl.yaml]: or is this the tag you talked about?

File: base/iam.tf (hunk 1 above)
Comment (duboisph) [reply base/iam.tf]: Right, missed that one, I'll make sure we've got only one 😓

**Comment by duboisph** [comment]
@SamClinckspoor 
> Would it be a lot of extra work to allow scaling on custom instancegroups?

No, since I've opted for autodiscovery, it's only a matter of Tagging the custom instancegroups with:

```
k8s.io/cluster-autoscaler/enabled
kubernetes.io/cluster/${cluster-name}
```

Some extra info to answer your above questions:
- It only checks if the cloud tags exist, the value is ignored.
- The `kubernetes.io/cluster/${cluster-name}` tag takes care of namespacing so staging / production in the same account don't start changing each other. Unfortunately it's not possible to do via IAM, since those ASG permissions act on '*'.

**Comment by iuriaranda** [comment]
@duboisph we could restrict the IAM policy to only apply to the correct autoscaling groups no?
For example: `*.${cluster_name}`

**Comment by duboisph** [comment]
@iuriaranda [Re-checked it in the AWS docs](https://docs.aws.amazon.com/autoscaling/ec2/userguide/control-access-using-iam.html) and looks like the cluster-autoscaler docs were wrong about IAM. The `Describe*` actions can only be applied to all resources, but looks like we should be able to limit `SetDesiredCapacity` and `TerminateInstanceInAutoScalingGroup`.

I'll make the change & test.

**Comment by duboisph** [comment]
IAM restrictions added & verified. Merging to master.

**Pull request merged**